File: config/layouts/grid.lua
local M = {}

M.name = "grid"
M.gap = 8
M.config = {}

function M.arrange(area, windows)
    local count = #windows
    if count == 0 then
        return {}
    end

    if count == 1 then
        local win = windows[1]
        win.x = area.x + M.gap
        win.y = area.y + M.gap
        win.width = area.width - M.gap * 2
        win.height = area.height - M.gap * 2
        return {win}
    end

    local cols = math.ceil(math.sqrt(count))
    local rows = math.ceil(count / cols)

    local cell_w = math.floor((area.width - M.gap * (cols + 1)) / cols)
    local cell_h = math.floor((area.height - M.gap * (rows + 1)) / rows)

    local result = {}

    for i = 1, count do
        local col = (i - 1) % cols
        local row = math.floor((i - 1) / cols)

        local win = windows[i]
        win.x = area.x + M.gap + col * (cell_w + M.gap)
        win.y = area.y + M.gap + row * (cell_h + M.gap)
        win.width = cell_w
        win.height = cell_h
        result[#result + 1] = win
    end

    return result
end

return M

File: core/src/layout/grid.rs
use crate::utils::Rect;
use crate::window::WindowId;

use super::Layout;

pub struct Grid {
    pub columns: usize,
    pub gap: i32,
    pub outer_gap: i32,
}

impl Default for Grid {
    fn default() -> Self {
        Grid {
            columns: 2,
            gap: 8,
            outer_gap: 16,
        }
    }
}

impl Layout for Grid {
    fn name(&self) -> &'static str {
        "grid"
    }

    fn arrange(&self, area: Rect, windows: &[WindowId]) -> Vec<Rect> {
        if windows.is_empty() {
            return Vec::new();
        }
        if windows.len() == 1 {
            return vec![area.grow(-self.outer_gap)];
        }
        let count = windows.len();
        let columns = self.columns.clamp(1, count.max(1));
        let rows = (count + columns - 1) / columns;
        let inner = area.grow(-self.outer_gap);
        let cell_width = (inner.width - self.gap * (columns as i32 - 1)) / columns as i32;
        let cell_height = (inner.height - self.gap * (rows as i32 - 1)) / rows as i32;
        let mut result = Vec::with_capacity(count);
        for index in 0..count {
            let row = index / columns;
            let col = index % columns;
            result.push(Rect::new(
                inner.x + col as i32 * (cell_width + self.gap),
                inner.y + row as i32 * (cell_height + self.gap),
                cell_width,
                cell_height,
            ));
        }
        result
    }
}

pub fn describe() -> &'static str {
    "grid layout with equally sized cells"
}